[4.6/4.7/4.8 Regression] Multiple copy constructors for template class lead to link errors

 http://gcc.gnu.org/bugzilla/show_bug.cgi?id=55249

--- gcc-4_6-branch/gcc/testsuite/g++.dg/template/array25.C	(revision 0)
+++ gcc-4_6-branch/gcc/testsuite/g++.dg/template/array25.C	(revision 194285)
@@ -0,0 +1,18 @@
+// PR c++/55249
+
+template <typename _Tp> struct A
+{
+  _Tp _M_instance[1];
+};
+template <class> struct inner_type
+{
+    inner_type () {}
+    inner_type (inner_type &);
+    inner_type (const inner_type &) {}
+};
+
+int
+main ()
+{
+    A <inner_type <int> > a, b = a;
+}
--- gcc-4_6-branch/gcc/cp/tree.c	(revision 194284)
+++ gcc-4_6-branch/gcc/cp/tree.c	(revision 194285)
@@ -490,7 +490,8 @@
   argvec = make_tree_vector ();
   if (init)
     {
-      tree dummy = build_dummy_object (inner_type);
+      tree init_type = strip_array_types (TREE_TYPE (init));
+      tree dummy = build_dummy_object (init_type);
       if (!real_lvalue_p (init))
 	dummy = move (dummy);
       VEC_quick_push (tree, argvec, dummy);
--- gcc-4_6-branch/gcc/cp/ChangeLog	(revision 194284)
+++ gcc-4_6-branch/gcc/cp/ChangeLog	(revision 194285)
@@ -1,3 +1,8 @@
+2012-12-06  Jason Merrill  <jason@redhat.com>
+
+	PR c++/55249
+	* tree.c (build_vec_init_elt): Use the type of the initializer.
+
 2012-11-29  Jason Merrill  <jason@redhat.com>
 
 	PR c++/51662
